File: python3/Linked Lists/linked_list_intersection.py
from ds import ListNode

"""
Definition of ListNode:
class ListNode:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
"""


def linked_list_intersection(head_A: ListNode, head_B: ListNode) -> ListNode:
    ptr_A, ptr_B = head_A, head_B
    # Traverse through list A with 'ptr_A' and list B with 'ptr_B' 
    # until they meet.
    while ptr_A != ptr_B:
        # Traverse list A -> list B by first traversing 'ptr_A' and 
        # then, upon reaching the end of list A, continue the 
        # traversal from the head of list B.
        ptr_A = ptr_A.next if ptr_A else head_B
        # Simultaneously, traverse list B -> list A.
        ptr_B = ptr_B.next if ptr_B else head_A
    # At this point, 'ptr_A' and 'ptr_B' either point to the 
    # intersection node or both are null if the lists do not 
    # intersect. Return either pointer.
    return ptr_A

File: java/Linked Lists/LinkedListIntersection.java
import DS.ListNode;

/*
    // Definition of ListNode:
    class ListNode {
        public int val;
        public ListNode next;
        public ListNode(int val) { this.val = val; }
        public ListNode(int val, ListNode next) {
            this(val);
            this.next = next;
        }
    }
 */


public class LinkedListIntersection {
    public static ListNode linkedListIntersection(ListNode headA, ListNode headB) {
        ListNode ptrA = headA;
        ListNode ptrB = headB;
        // Traverse through list A with 'ptrA' and list B with 'ptrB'
        // until they meet.
        while (ptrA != ptrB) {
            // Traverse list A -> list B by first traversing 'ptrA' and
            // then, upon reaching the end of list A, continue the
            // traversal from the head of list B.
            ptrA = ptrA == null ? headB : ptrA.next;
            // Simultaneously, traverse list B -> list A.
            ptrB = ptrB == null ? headA : ptrB.next;
        }
        // At this point, 'ptrA' and 'ptrB' either point to the
        // intersection node or both are null if the lists do not
        // intersect. Return either pointer.
        return ptrA;
    }
}
